resource-aggregate/cqrs/eventstore/mongodb/eventstore.go

```diff
@@ -58,68 +91,706 @@ func NewEventStore(config Config, goroutinePoolGo eventstore.GoroutinePoolGoFunc
+				errors = append(errors, fmt.Errorf("cannot load events to eventstore model: %w", err))
+			}
+			queries = queries[:0]
+		}
+	}
+	if len(errors) > 0 {
+		return fmt.Errorf("loader cannot load events: %v", errors)
+	}
+
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+
+	if len(queries) > 0 {
+		return l.store.LoadFromVersion(ctx, queries, l.eventHandler)
+	}
+
+	return nil
+}
+
+func (l *loader) QueryHandlePool(ctx context.Context, iter *queryIterator) error {
+	var query eventstore.VersionQuery
+	queries := make([]eventstore.VersionQuery, 0, 128)
+	var wg sync.WaitGroup
+
+	var errors []error
+	var errorsLock sync.Mutex
+
+	for iter.Next(ctx, &query) {
+		queries = append(queries, query)
+		if len(queries) >= l.store.batchSize {
+			wg.Add(1)
+			l.store.LogDebugfFunc("mongodb:loader:QueryHandlePool:newTask")
+			tmp := queries
+			err := l.store.goroutinePoolGo(func() {
+				defer wg.Done()
+				l.store.LogDebugfFunc("mongodb:loader:QueryHandlePool:task:LoadFromVersion:start")
+				err := l.store.LoadFromVersion(ctx, tmp, l.eventHandler)
+				l.store.LogDebugfFunc("mongodb:loader:QueryHandlePool:task:LoadFromVersion:done")
+				if err != nil {
+					errorsLock.Lock()
+					defer errorsLock.Unlock()
+					errors = append(errors, fmt.Errorf("cannot load events to eventstore model: %w", err))
+				}
+				l.store.LogDebugfFunc("mongodb:loader:QueryHandlePool:doneTask")
+			})
+			if err != nil {
+				wg.Done()
+				errorsLock.Lock()
+				errors = append(errors, fmt.Errorf("cannot submit task to load events to eventstore model: %w", err))
+				errorsLock.Unlock()
+				break
+			}
+			queries = make([]eventstore.VersionQuery, 0, 128)
+		}
+	}
+	wg.Wait()
+	if len(errors) > 0 {
+		return fmt.Errorf("loader cannot load events: %v", errors)
+	}
+
+	if iter.Err() != nil {
+		return iter.Err()
+	}
+	if len(queries) > 0 {
+		return l.store.LoadFromVersion(ctx, queries, l.eventHandler)
+	}
+
+	return nil
+}
+
+func (s *EventStore) loadEvents(ctx context.Context, queries []eventstore.VersionQuery, eh eventstore.Handler, funcToMgoQuery func(queries []eventstore.VersionQuery) (primitive.M, error)) error {
+	collections := make(map[string][]eventstore.VersionQuery)
+	for _, query := range queries {
+		collections[query.GroupID] = append(collections[query.GroupID], query)
+	}
+
+	var errors []error
+	for groupID, queries := range collections {
+		q, err := funcToMgoQuery(queries)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("cannot load events version: %w", err))
+			continue
+		}
+		err = s.loadMgoQuery(ctx, groupID, eh, q)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+	}
+	if len(errors) > 0 {
+		return fmt.Errorf("%+v", errors)
+	}
+	return nil
+}
+
+// LoadUpToVersion loads aggragates events up to a specific version.
+func (s *EventStore) LoadUpToVersion(ctx context.Context, queries []eventstore.VersionQuery, eh eventstore.Handler) error {
+	s.LogDebugfFunc("mongodb.Eventstore.LoadUpToVersion start")
```

Comment:
We should get rid of these logs. Probably not yet, but definitely before merging.